// @asymmetrik/ngx-leaflet-markercluster/dist/leaflet-markercluster/leaflet-markercluster.module.js
import * as ɵngcc0 from '@angular/core';
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { NgModule } from '@angular/core';
import { LeafletMarkerClusterDirective } from './leaflet-markercluster.directive';
var LeafletMarkerClusterModule = /** @class */ (function () {
    function LeafletMarkerClusterModule() {
    }
    LeafletMarkerClusterModule_1 = LeafletMarkerClusterModule;
    LeafletMarkerClusterModule.forRoot = function () {
        return { ngModule: LeafletMarkerClusterModule_1, providers: [] };
    };
    var LeafletMarkerClusterModule_1;
LeafletMarkerClusterModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: LeafletMarkerClusterModule });
LeafletMarkerClusterModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function LeafletMarkerClusterModule_Factory(t) { return new (t || LeafletMarkerClusterModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LeafletMarkerClusterModule, { declarations: function () { return [LeafletMarkerClusterDirective]; }, exports: function () { return [LeafletMarkerClusterDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LeafletMarkerClusterModule, [{
        type: NgModule,
        args: [{
                exports: [LeafletMarkerClusterDirective],
                declarations: [LeafletMarkerClusterDirective]
            }]
    }], function () { return []; }, null); })();
    return LeafletMarkerClusterModule;
}());
export { LeafletMarkerClusterModule };
