// devextreme-angular/esm5/ui/nested/item-view.js
/*!
 * devextreme-angular
 * Version: 19.2.6
 * Build date: Thu Jan 30 2020
 *
 * Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
import * as tslib_1 from "tslib";
/* tslint:disable:max-line-length */
import { Component, NgModule, Host, SkipSelf, Input } from '@angular/core';
import { NestedOptionHost, } from 'devextreme-angular/core';
import { NestedOption } from 'devextreme-angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'devextreme-angular/core';
var DxoItemViewComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DxoItemViewComponent, _super);
    function DxoItemViewComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxoItemViewComponent.prototype, "mode", {
        get: function () {
            return this._getOption('mode');
        },
        set: function (value) {
            this._setOption('mode', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoItemViewComponent.prototype, "showFolders", {
        get: function () {
            return this._getOption('showFolders');
        },
        set: function (value) {
            this._setOption('showFolders', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoItemViewComponent.prototype, "showParentFolder", {
        get: function () {
            return this._getOption('showParentFolder');
        },
        set: function (value) {
            this._setOption('showParentFolder', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoItemViewComponent.prototype, "_optionPath", {
        get: function () {
            return 'itemView';
        },
        enumerable: true,
        configurable: true
    });
    DxoItemViewComponent.prototype.ngOnDestroy = function () {
        this._addRemovedOption(this._fullOptionPath().slice(0, -1));
    };
    DxoItemViewComponent.ctorParameters = function () { return [
        { type: NestedOptionHost, decorators: [{ type: SkipSelf }, { type: Host }] },
        { type: NestedOptionHost, decorators: [{ type: Host }] }
    ]; };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", String),
        tslib_1.__metadata("design:paramtypes", [String])
    ], DxoItemViewComponent.prototype, "mode", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], DxoItemViewComponent.prototype, "showFolders", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], DxoItemViewComponent.prototype, "showParentFolder", null);
    DxoItemViewComponent = tslib_1.__decorate([ tslib_1.__param(0, SkipSelf()), tslib_1.__param(0, Host()),
        tslib_1.__param(1, Host()),
        tslib_1.__metadata("design:paramtypes", [NestedOptionHost,
            NestedOptionHost])
    ], DxoItemViewComponent);
DxoItemViewComponent.ɵfac = function DxoItemViewComponent_Factory(t) { return new (t || DxoItemViewComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NestedOptionHost, 5), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NestedOptionHost, 1)); };
DxoItemViewComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: DxoItemViewComponent, selectors: [["dxo-item-view"]], inputs: { mode: "mode", showFolders: "showFolders", showParentFolder: "showParentFolder" }, features: [ɵngcc0.ɵɵProvidersFeature([NestedOptionHost]), ɵngcc0.ɵɵInheritDefinitionFeature], decls: 0, vars: 0, template: function DxoItemViewComponent_Template(rf, ctx) { }, styles: [""] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DxoItemViewComponent, [{
        type: Component,
        args: [{
                selector: 'dxo-item-view',
                template: '',
                providers: [NestedOptionHost],
                styles: ['']
            }]
    }], function () { return [{ type: ɵngcc1.NestedOptionHost, decorators: [{
                type: SkipSelf
            }, {
                type: Host
            }] }, { type: ɵngcc1.NestedOptionHost, decorators: [{
                type: Host
            }] }]; }, { mode: [{
            type: Input
        }], showFolders: [{
            type: Input
        }], showParentFolder: [{
            type: Input
        }] }); })();
    return DxoItemViewComponent;
}(NestedOption));
export { DxoItemViewComponent };
var DxoItemViewModule = /** @class */ (function () {
    function DxoItemViewModule() {
    }
DxoItemViewModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: DxoItemViewModule });
DxoItemViewModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function DxoItemViewModule_Factory(t) { return new (t || DxoItemViewModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DxoItemViewModule, { declarations: [DxoItemViewComponent], exports: [DxoItemViewComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DxoItemViewModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    DxoItemViewComponent
                ],
                exports: [
                    DxoItemViewComponent
                ]
            }]
    }], function () { return []; }, null); })();
    return DxoItemViewModule;
}());
export { DxoItemViewModule };
